File: math/oddMagicSquare.pl
#!/usr/bin/perl -w

# Written by Bill Brassfield in 2015
# This program generates odd-ordered magic squares (3x3 through 31x31)

use strict;

my $size = $ARGV[0];
if (!defined($size) || ($size !~/^\d+$/) || ($size < 3) || ($size > 31) || ($size % 2 == 0))
{
  die "\n  Usage: $0 <odd-number-from-3-to-31>\n\n";
}

my $msq  = [];
for (my $i=0; $i<$size; $i++)
{
  my $row = [];
  for (my $j=0; $j<$size; $j++)
  {
    push(@$row, undef);
  }
  push(@$msq, $row);
}

my $xPos = ($size - 1) / 2;
my $yPos = 0;
for (my $i=1; $i<=$size*$size; $i++)
{
  $msq->[$yPos]->[$xPos] = $i;

  $xPos++;
  $yPos--;

  if (($xPos >= $size) && ($yPos < 0))
  {
    $xPos--;
    $yPos++;  $yPos++;
  }
  elsif (($xPos >= $size) && ($yPos >= 0) && ($yPos < $size))
  {
    $xPos = 0;
  }
  elsif (($yPos < 0) && ($xPos >= 0) && ($xPos < $size))
  {
    $yPos = $size - 1;
  }
  elsif (defined($msq->[$yPos]->[$xPos]))
  {
    $xPos--;
    $yPos++;  $yPos++;
  }
}

for ($yPos=0; $yPos<$size; $yPos++)
{
  for ($xPos=0; $xPos<$size; $xPos++)
  {
    printf(" %3d ", $msq->[$yPos]->[$xPos]);
  }
  print "\n";
}

File: math/oddMagicSquare.rb
#!/usr/bin/ruby

# Written by Bill Brassfield in 2015
# This program generates odd-ordered magic squares (3x3 through 31x31)

size = Integer(ARGV[0])

if ((size < 3) || (size > 31) || (size % 2 == 0))
  print "\n  Usage: <odd-number-from-3-to-31>\n\n"
  exit(1)
end

msq = Array.new(size, 0)
0.upto(size-1) { |i|
  msq[i] = Array.new(size, 0)
  0.upto(size-1) { |j|
    msq[i][j] = 0
  }
}

xpos = (size - 1) / 2
ypos = 0

1.upto(size*size) { |i|
  msq[ypos][xpos] = i

  xpos += 1;  ypos -= 1

  if ((xpos >= size) && (ypos < 0))
    xpos -= 1;  ypos += 2
  elsif ((xpos >= size) && (ypos >= 0) && (ypos < size))
    xpos = 0
  elsif ((ypos < 0) && (xpos >= 0) && (xpos < size))
    ypos = size - 1
  elsif (msq[ypos][xpos] != 0)
    xpos -= 1;  ypos += 2
  end
}


0.upto(size-1) { |i|
  0.upto(size-1) { |j|
    printf(" %3d ", msq[i][j])
  }
  puts
}
